File: src/venues.ts
import { readFileSync } from "node:fs";
import { dirname, resolve } from "node:path";
import { fileURLToPath } from "node:url";

export type VenueRecord = {
  name: string;
  full_name: string;
  ccf_category: string;
  domain: string;
  venue_type: string;
  eligible_tracks: string[];
  ineligible_tracks: string[];
  source_url: string;
  dblp_url: string;
  last_checked: string;
  provenance_note: string;
};

export type DomainProfile = {
  key: string;
  label: string;
  primary_venues: string[];
  secondary_venues: string[];
  review_focus: string[];
  keywords: string[];
  aliases: string[];
  venue_records: Record<string, VenueRecord>;
};

export type VenueDatabase = {
  version: string;
  source_note: string;
  domains: Record<string, DomainProfile>;
};

export type DomainRoute = {
  domain: DomainProfile;
  score: number;
  matched_keywords: string[];
  requested: boolean;
};

export function loadVenueDatabase(path?: string): VenueDatabase {
  const dataPath =
    path ??
    resolve(dirname(fileURLToPath(import.meta.url)), "..", "data", "venues.json");
  const raw = JSON.parse(readFileSync(dataPath, "utf8").replace(/^\uFEFF/, "")) as {
    version: string;
    source_note: string;
    domains: Record<string, Omit<DomainProfile, "key" | "venue_records"> & { venue_records?: VenueRecord[] }>;
  };
  const domains: Record<string, DomainProfile> = {};
  for (const [key, value] of Object.entries(raw.domains)) {
    const records: Record<string, VenueRecord> = {};
    for (const record of value.venue_records ?? []) records[record.name] = record;
    domains[key] = {
      key,
      label: value.label,
      primary_venues: value.primary_venues,
      secondary_venues: value.secondary_venues,
      review_focus: value.review_focus,
      keywords: value.keywords,
      aliases: value.aliases ?? [],
      venue_records: records
    };
  }
  return { version: raw.version, source_note: raw.source_note, domains };
}

export function routeIdea(
  idea: string,
  database = loadVenueDatabase(),
  requestedDomains: string[] = []
): DomainRoute[] {
  const normalized = idea.toLocaleLowerCase();
  const requested = new Set(requestedDomains.map(normalizeRequest));
  const routes: DomainRoute[] = [];
  for (const [key, profile] of Object.entries(database.domains)) {
    const matched = profile.keywords.filter((keyword) => normalized.includes(keyword.toLocaleLowerCase()));
    const candidates = new Set([
      normalizeRequest(key),
      normalizeRequest(profile.label),
      ...profile.aliases.map(normalizeRequest),
      ...profile.primary_venues.map(normalizeRequest),
      ...profile.secondary_venues.map(normalizeRequest)
    ]);
    const requestedMatch = [...requested].some((value) => candidates.has(value));
    routes.push({
      domain: profile,
      score: matched.length * 10 + (requestedMatch ? 100 : 0),
      matched_keywords: matched,
      requested: requestedMatch
    });
  }
  routes.sort((a, b) => Number(b.requested) - Number(a.requested) || b.score - a.score || b.domain.label.localeCompare(a.domain.label));
  if (routes[0]?.score) return routes;
  const ai = database.domains.ai_llm_agent;
  if (!ai) return routes;
  return [{ domain: ai, score: 1, matched_keywords: [], requested: false }, ...routes.filter((route) => route.domain.key !== "ai_llm_agent")];
}

export function validateVenueDatabase(database = loadVenueDatabase()): string[] {
  const errors: string[] = [];
  if (!database.version) errors.push("database version is required");
  for (const [key, profile] of Object.entries(database.domains)) {
    const listed = new Set([...profile.primary_venues, ...profile.secondary_venues]);
    for (const venue of listed) {
      if (!profile.venue_records[venue]) errors.push(`${key}: missing venue record for ${venue}`);
    }
    for (const [name, record] of Object.entries(profile.venue_records)) {
      if (record.domain !== key) errors.push(`${key}: ${name} has mismatched domain ${record.domain}`);
      if (!["A", "B", "C"].includes(record.ccf_category)) errors.push(`${key}: ${name} has invalid CCF category ${record.ccf_category}`);
      if (!["conference", "journal"].includes(record.venue_type)) errors.push(`${key}: ${name} has invalid venue type ${record.venue_type}`);
      if (!/^https?:\/\//.test(record.source_url)) errors.push(`${key}: ${name} source_url must be absolute`);
      if (!/^https?:\/\//.test(record.dblp_url)) errors.push(`${key}: ${name} dblp_url must be absolute`);
      if (!record.eligible_tracks.includes("Full paper") && !record.eligible_tracks.includes("Regular paper")) {
        errors.push(`${key}: ${name} must include Full paper or Regular paper eligibility`);
      }
      for (const track of ["Workshop", "Demo", "Short paper"]) {
        if (!record.ineligible_tracks.includes(track)) errors.push(`${key}: ${name} must mark workshop/demo/short paper as ineligible`);
      }
    }
  }
  return errors;
}

function normalizeRequest(value: string): string {
  return value.toLocaleLowerCase().replace(/[_\-\/\\|,:;]/g, " ").split(/\s+/).filter(Boolean).join(" ");
}

File: idea2repo/venues.py
"""Venue database loading and domain routing."""

from __future__ import annotations

import json
from dataclasses import dataclass
from importlib import resources
from pathlib import Path
from typing import Any


@dataclass(frozen=True)
class VenueRecord:
    """Auditable metadata for one venue seed entry."""

    name: str
    full_name: str
    ccf_category: str
    domain: str
    venue_type: str
    eligible_tracks: tuple[str, ...]
    ineligible_tracks: tuple[str, ...]
    source_url: str
    dblp_url: str
    last_checked: str
    provenance_note: str


@dataclass(frozen=True)
class DomainProfile:
    """A review-style profile for one target research domain."""

    key: str
    label: str
    primary_venues: tuple[str, ...]
    secondary_venues: tuple[str, ...]
    review_focus: tuple[str, ...]
    keywords: tuple[str, ...]
    aliases: tuple[str, ...]
    venue_records: dict[str, VenueRecord]


@dataclass(frozen=True)
class VenueDatabase:
    """Updatable venue and reviewer rubric database."""

    version: str
    source_note: str
    domains: dict[str, DomainProfile]


@dataclass(frozen=True)
class DomainRoute:
    """Domain routing result for a raw idea."""

    domain: DomainProfile
    score: int
    matched_keywords: tuple[str, ...]
    requested: bool = False


def load_venue_database(path: str | Path | None = None) -> VenueDatabase:
    """Load the venue database from package data or a user-supplied JSON file."""

    if path is None:
        data = resources.files("idea2repo").joinpath("data/venues.json").read_text(
            encoding="utf-8-sig"
        )
    else:
        data = Path(path).read_text(encoding="utf-8-sig")

    raw: dict[str, Any] = json.loads(data)
    domains = {
        key: DomainProfile(
            key=key,
            label=value["label"],
            primary_venues=tuple(value["primary_venues"]),
            secondary_venues=tuple(value["secondary_venues"]),
            review_focus=tuple(value["review_focus"]),
            keywords=tuple(value["keywords"]),
            aliases=tuple(value.get("aliases", ())),
            venue_records=_load_venue_records(key, value.get("venue_records", [])),
        )
        for key, value in raw["domains"].items()
    }
    return VenueDatabase(
        version=raw["version"],
        source_note=raw["source_note"],
        domains=domains,
    )


def route_idea(
    idea: str,
    database: VenueDatabase | None = None,
    *,
    requested_domains: list[str] | None = None,
) -> list[DomainRoute]:
    """Rank target domains from keyword evidence and optional user choices."""

    database = database or load_venue_database()
    normalized = idea.casefold()
    requested = {_normalize_request(value) for value in requested_domains or []}
    routes: list[DomainRoute] = []

    for key, profile in database.domains.items():
        matches = tuple(
            keyword for keyword in profile.keywords if keyword.casefold() in normalized
        )
        score = len(matches) * 10
        candidates = {
            _normalize_request(key),
            _normalize_request(profile.label),
        }
        candidates.update(_normalize_request(value) for value in profile.aliases)
        candidates.update(
            _normalize_request(value)
            for value in profile.primary_venues + profile.secondary_venues
        )
        requested_match = bool(requested & candidates)
        if requested_match:
            score += 100
        routes.append(DomainRoute(profile, score, matches, requested_match))

    routes.sort(key=lambda route: (route.requested, route.score, route.domain.label), reverse=True)
    if routes and routes[0].score > 0:
        return routes

    return [
        DomainRoute(
            database.domains["ai_llm_agent"],
            1,
            (),
        )
    ] + [
        route
        for route in routes
        if route.domain.key != "ai_llm_agent"
    ]


def _normalize_request(value: str) -> str:
    separators = str.maketrans({char: " " for char in "_-/\\|,:;"})
    return " ".join(value.casefold().translate(separators).split())


def validate_venue_database(database: VenueDatabase | None = None) -> tuple[str, ...]:
    """Validate venue provenance and CCF track metadata."""

    database = database or load_venue_database()
    errors: list[str] = []
    if not database.version:
        errors.append("database version is required")
    for key, profile in database.domains.items():
        listed = set(profile.primary_venues + profile.secondary_venues)
        missing = sorted(listed - set(profile.venue_records))
        for venue in missing:
            errors.append(f"{key}: missing venue record for {venue}")
        for name, record in profile.venue_records.items():
            if record.domain != key:
                errors.append(f"{key}: {name} has mismatched domain {record.domain}")
            if record.ccf_category not in {"A", "B", "C"}:
                errors.append(f"{key}: {name} has invalid CCF category {record.ccf_category}")
            if record.venue_type not in {"conference", "journal"}:
                errors.append(f"{key}: {name} has invalid venue type {record.venue_type}")
            if not record.source_url.startswith(("https://", "http://")):
                errors.append(f"{key}: {name} source_url must be absolute")
            if not record.dblp_url.startswith(("https://", "http://")):
                errors.append(f"{key}: {name} dblp_url must be absolute")
            if "Full paper" not in record.eligible_tracks and "Regular paper" not in record.eligible_tracks:
                errors.append(f"{key}: {name} must include Full paper or Regular paper eligibility")
            required_ineligible = {"Workshop", "Demo", "Short paper"}
            if not required_ineligible.issubset(set(record.ineligible_tracks)):
                errors.append(f"{key}: {name} must mark workshop/demo/short paper as ineligible")
            if set(record.eligible_tracks) & set(record.ineligible_tracks):
                errors.append(f"{key}: {name} has contradictory eligible and ineligible tracks")
            if required_ineligible & set(record.eligible_tracks):
                errors.append(f"{key}: {name} cannot mark workshop/demo/short paper as eligible")
            if not record.last_checked:
                errors.append(f"{key}: {name} last_checked is required")
    return tuple(errors)


def _load_venue_records(domain: str, records: list[dict[str, Any]]) -> dict[str, VenueRecord]:
    loaded: dict[str, VenueRecord] = {}
    for value in records:
        required = (
            "name",
            "full_name",
            "ccf_category",
            "domain",
            "venue_type",
            "eligible_tracks",
            "ineligible_tracks",
            "source_url",
            "dblp_url",
            "last_checked",
            "provenance_note",
        )
        missing = [field for field in required if field not in value]
        if missing:
            name = value.get("name", "<unknown>")
            raise ValueError(f"venue record {name} missing required fields: {', '.join(missing)}")
        record = VenueRecord(
            name=value["name"],
            full_name=value["full_name"],
            ccf_category=value["ccf_category"],
            domain=value["domain"],
            venue_type=value["venue_type"],
            eligible_tracks=tuple(value["eligible_tracks"]),
            ineligible_tracks=tuple(value["ineligible_tracks"]),
            source_url=value["source_url"],
            dblp_url=value["dblp_url"],
            last_checked=value["last_checked"],
            provenance_note=value["provenance_note"],
        )
        loaded[record.name] = record
    return loaded
